truffle/src/com.oracle.truffle.runtime/src/com/oracle/truffle/runtime/OptimizedCallTarget.java

```diff
@@ -855,6 +858,60 @@ final boolean isCompilationFailed() {
+        long sct = runtime().stoppedCompilationTime().get();
+        if (compilationActivityMode == OptimizedTruffleRuntime.CompilationActivityMode.STOP_COMPILATION) {
+            if (sct != 0 && System.currentTimeMillis() - sct > engine.stoppedCompilationRetryDelay) {
+                runtime().stoppedCompilationTime().compareAndSet(sct, 0);
+                // Try again every StoppedCompilationRetryDelay milliseconds to potentially trigger a code cache sweep.
+                compilationActivityMode = OptimizedTruffleRuntime.CompilationActivityMode.RUN_COMPILATION;
+            }
+        }
+
+        switch (compilationActivityMode) {
+            case RUN_COMPILATION : {
+                // This is the common case - compilations are not stopped.
+                return false;
+            }
+            case STOP_COMPILATION : {
+                if (sct == 0) {
+                    runtime().stoppedCompilationTime().compareAndSet(0, System.currentTimeMillis());
+                }
+                // Flush the compilations queue for now. There's still a chance that compilation
+                // will be re-enabled eventually, if the hosts code cache can be cleaned up.
+                Collection<OptimizedCallTarget> targets = runtime().getCompileQueue().getQueuedTargets(null);
+                // If there's just a single compilation target in the queue, the chance is high that it is
+                // the one we've just added after the StoppedCompilationRetryDelay ran out, so keep it to
+                // potentially trigger a code cache sweep.
+                if (targets.size() > 1) {
+                    for (OptimizedCallTarget target : targets) {
+                        target.cancelCompilation("Compilation temporary disabled due to full code cache.");
+                    }
+                }
+                return true;
+            }
+            case SHUTDOWN_COMPILATION : {
+                // Compilation was shut down permanently because the hosts code cache ran full and
+                // the host was configured without support for code cache sweeping.
+                TruffleLogger logger = engine.getLogger("engine");
+                // The logger can be null if the engine is closed.
+                if (logger != null && runtime().logShutdownCompilations().compareAndExchange(true, false)) {
+                    logger.log(Level.WARNING, "Truffle compilations permanently disabled because of full code cache. "  +
+                            "Increase the code cache size using '-XX:ReservedCodeCacheSize=' and/or run with '-XX:+UseCodeCacheFlushing -XX:+MethodFlushing'.");
+                }
+                try {
+                    runtime().getCompileQueue().shutdownAndAwaitTermination(100 /* milliseconds */);
+                } catch (RuntimeException re) {
+                    // Best effort, ignore failure
+                }
+                return true;
```

Comment:
Here is my suggestion:

Note instead of runtime we use the engine to log to avoid missing warnings. (warnings can be suppressed per engine)


```java
      if (engine.logShutdownCompilations().get()) {
                // The logger can be null if the engine is closed.
                if (engine.logShutdownCompilations().compareAndExchange(true, false)) {
                       TruffleLogger logger = engine.getLogger("engine");
                       if (logger != null) {
                          // log for each engine once
                      }
                }
      }
      // do not try to compile this call target again
      callTarget.compilationFailed = true;
      return true;
```

Reply:
On a second thought, it might be a good idea to return the enum here, so the compilationFailed flag can be set in OptimizedCallTarget.compile and not in a method called `isCompilationStopped` which is a bit surprising.